### PR TITLE
177618304 epoch wall clock

### PR DESCRIPTION
**Pull request by maciejkocylapc**
#### What's this PR do?
Create EpochWallClock model among with api and data gathering

#### How should this be manually tested?
After running gather_rpc_data.rb and gather_rpc_data_mainnet.rb scripts there should be new record in db. 
API endpoints should work according to documentation.

#### What are the relevant tickets?
https://www.pivotaltracker.com/story/show/177618304

#### Task completed checklist:
- [Y] Is there appropriate test coverage?
- [Y] Did I take into consideration possible security vulnerabilities?
- [Y] Are the controllers secure?
- [Y] Was every important matter documented?

#### Things to watch out for
- [Y] Does this PR have any migrations?
- [N] Does this PR add new dependencies?

### CODE REVIEW

**Comment by brianlong** [comment]
@maciejkocylapc Please include the ending_slot too. If you can do this tomorrow morning your time, I will merge & deploy tomorrow my time.

**Review by tvannurden — CHANGES_REQUESTED** [review]

File: app/logic/solana_logic.rb [hunk 1]
```diff
@@ -52,6 +52,27 @@ def epoch_get
         slot_index: epoch_json['slotIndex'],
         slots_in_epoch: epoch_json['slotsInEpoch']
       )
+      begin
+        EpochWallClock.create!(
```
Comment (maciejkocylapc) [inline app/logic/solana_logic.rb:56]: I did some simulations and the speed is very similar to raw INSERT IGNORE query. In my opinion this solution is a bit more secure.

**Comment by kbiala** [comment]
@tvannurden Looks like you've already started CR in this story. Can you take another looks and accept if ready? I deployed this branch to staging.

**Comment by tvannurden** [comment]
@maciejkocylapc you're just missing some API documentation and tests for the pagination. What is the default number of results returned? What is the max?

File: app/views/public/api_documentation.html.erb [hunk 2]
```diff
@@ -208,6 +208,11 @@
   <div class="col-sm-12">
     <p>The Epoch endpoint will return all epochs data. Replace `:network` with 'testnet' or 'mainnet' as desired. </p>
     <p>Request:</p>
+    <pre>curl -H "Token: secret-api-token" 'https://www.validators.app/api/v1/epoch/:network/index.json'</pre>
```
Comment (tvannurden) [inline app/views/public/api_documentation.html.erb]: I made some minor improvements and that is no longer the endpoint, please check `rails routes`

File: app/views/public/api_documentation.html.erb [hunk 3]
```diff
@@ -208,6 +208,11 @@
...
+    <p>Parameter Options:</p>
+    <p>`per=NN` number of records per page, defaults to 50</p>
+    <pre>curl -H "Token: secret-api-token" 'https://www.validators.app/api/v1/epoch/:network/index.json?per=50'</pre>
```
Comment (tvannurden) [inline app/views/public/api_documentation.html.erb]: Same with here

File: test/controllers/api/v1/epochs_controller_test.rb [hunk 4]
```diff
@@ -40,4 +40,30 @@ def setup
       created_at
     ].sort, resp[0].keys.sort
   end
+
+  test 'epoch with default pagination returns correct number of records' do
+    100.times do |i|
```
Comment (tvannurden) [inline test/controllers/api/v1/epochs_controller_test.rb:45]: 51 times is probably fine for this one, and still tests what you're looking for

File: test/controllers/api/v1/epochs_controller_test.rb [hunk 5]
```diff
@@ -40,4 +40,30 @@ def setup
...
+
+    get api_v1_epoch_index_url(network: 'testnet'), headers: { 'Token' => @user.api_token }
+    resp = response_to_json(@response.body)
+
+    assert_response 200
+    assert_equal 50, resp.count
+  end
+
+  test 'epoch with custom pagination returns correct number of records' do
+    100.times do |i|
```
Comment (tvannurden) [inline test/controllers/api/v1/epochs_controller_test.rb:57]: Same with here, you don't need 100 records to test what yo'ure trying to test

**Comment by tvannurden** [comment]
What are all these other changes that leaked in?

**Comment by tvannurden** [comment]
I don't see any tests for the `total` key that you're returning in the API now

**Comment by tvannurden** [comment]
Also cannot run the automated tests anymore:
```
~/Dev/www.validators{177618304-epoch-wall-clock}$ rails test
warning: parser/current is loading parser/ruby26, which recognizes
warning: 2.6.6-compliant syntax, but you are running 2.6.5.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
/Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/message_encryptor.rb:203:in `rescue in _decrypt': ActiveSupport::MessageEncryptor::InvalidMessage (ActiveSupport::MessageEncryptor::InvalidMessage)
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/message_encryptor.rb:180:in `_decrypt'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/message_encryptor.rb:154:in `decrypt_and_verify'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/messages/rotator.rb:22:in `decrypt_and_verify'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/encrypted_file.rb:92:in `decrypt'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/encrypted_file.rb:54:in `read'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/encrypted_configuration.rb:21:in `read'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/encrypted_configuration.rb:33:in `config'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/encrypted_configuration.rb:38:in `options'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/core_ext/module/delegation.rb:309:in `method_missing'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/devise-4.7.3/lib/devise/secret_key_finder.rb:24:in `key_exists?'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/devise-4.7.3/lib/devise/secret_key_finder.rb:10:in `find'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/devise-4.7.3/lib/devise/rails.rb:37:in `block in <class:Engine>'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/railties-6.1.3/lib/rails/initializable.rb:32:in `instance_exec'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/railties-6.1.3/lib/rails/initializable.rb:32:in `run'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/railties-6.1.3/lib/rails/initializable.rb:61:in `block in run_initializers'
	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:228:in `block in tsort_each'
	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:431:in `each_strongly_connected_component_from'
	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:349:in `block in each_strongly_connected_component'
	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:347:in `each'
	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:347:in `call'
	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:347:in `each_strongly_connected_component'
	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:226:in `tsort_each'
	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:205:in `tsort_each'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/railties-6.1.3/lib/rails/initializable.rb:60:in `run_initializers'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/railties-6.1.3/lib/rails/application.rb:384:in `initialize!'
	from /Users/tylervannurden/Dev/www.validators/config/environment.rb:5:in `<main>'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/bootsnap-1.7.2/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/bootsnap-1.7.2/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/bootsnap-1.7.2/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/bootsnap-1.7.2/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/bootsnap-1.7.2/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/zeitwerk-2.4.2/lib/zeitwerk/kernel.rb:34:in `require'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/dependencies.rb:332:in `block in require'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/dependencies.rb:299:in `load_dependency'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/dependencies.rb:332:in `require'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application.rb:106:in `preload'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application.rb:157:in `serve'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application.rb:145:in `block in run'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application.rb:139:in `loop'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application.rb:139:in `run'
	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application/boot.rb:19:in `<top (required)>'
	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/site_ruby/2.6.0/rubygems/core_ext/kernel_require.rb:72:in `require'
	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/site_ruby/2.6.0/rubygems/core_ext/kernel_require.rb:72:in `require'
	from -e:1:in `<main>'
```

File: app/views/public/api_documentation.html.erb (hunk 2 above)
Comment (maciejkocylapc) [reply app/views/public/api_documentation.html.erb]: I've already fixed this in previous commit.

**Comment by maciejkocylapc** [comment]
> What are all these other changes that leaked in? The model annotations
. 
This is because of changing from utf8 to utf8mb4 in database.yml. It was explained on slack and has it's own PR

**Comment by maciejkocylapc** [comment]
> Also cannot run the automated tests anymore:
> 
> ```
> ~/Dev/www.validators{177618304-epoch-wall-clock}$ rails test
> warning: parser/current is loading parser/ruby26, which recognizes
> warning: 2.6.6-compliant syntax, but you are running 2.6.5.
> warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
> /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/message_encryptor.rb:203:in `rescue in _decrypt': ActiveSupport::MessageEncryptor::InvalidMessage (ActiveSupport::MessageEncryptor::InvalidMessage)
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/message_encryptor.rb:180:in `_decrypt'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/message_encryptor.rb:154:in `decrypt_and_verify'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/messages/rotator.rb:22:in `decrypt_and_verify'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/encrypted_file.rb:92:in `decrypt'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/encrypted_file.rb:54:in `read'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/encrypted_configuration.rb:21:in `read'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/encrypted_configuration.rb:33:in `config'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/encrypted_configuration.rb:38:in `options'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/core_ext/module/delegation.rb:309:in `method_missing'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/devise-4.7.3/lib/devise/secret_key_finder.rb:24:in `key_exists?'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/devise-4.7.3/lib/devise/secret_key_finder.rb:10:in `find'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/devise-4.7.3/lib/devise/rails.rb:37:in `block in <class:Engine>'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/railties-6.1.3/lib/rails/initializable.rb:32:in `instance_exec'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/railties-6.1.3/lib/rails/initializable.rb:32:in `run'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/railties-6.1.3/lib/rails/initializable.rb:61:in `block in run_initializers'
> 	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:228:in `block in tsort_each'
> 	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
> 	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:431:in `each_strongly_connected_component_from'
> 	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:349:in `block in each_strongly_connected_component'
> 	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:347:in `each'
> 	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:347:in `call'
> 	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:347:in `each_strongly_connected_component'
> 	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:226:in `tsort_each'
> 	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/tsort.rb:205:in `tsort_each'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/railties-6.1.3/lib/rails/initializable.rb:60:in `run_initializers'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/railties-6.1.3/lib/rails/application.rb:384:in `initialize!'
> 	from /Users/tylervannurden/Dev/www.validators/config/environment.rb:5:in `<main>'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/bootsnap-1.7.2/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/bootsnap-1.7.2/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/bootsnap-1.7.2/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/bootsnap-1.7.2/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/bootsnap-1.7.2/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/zeitwerk-2.4.2/lib/zeitwerk/kernel.rb:34:in `require'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/dependencies.rb:332:in `block in require'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/dependencies.rb:299:in `load_dependency'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/activesupport-6.1.3/lib/active_support/dependencies.rb:332:in `require'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application.rb:106:in `preload'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application.rb:157:in `serve'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application.rb:145:in `block in run'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application.rb:139:in `loop'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application.rb:139:in `run'
> 	from /Users/tylervannurden/.rvm/gems/ruby-2.6.5@validators/gems/spring-2.1.1/lib/spring/application/boot.rb:19:in `<top (required)>'
> 	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/site_ruby/2.6.0/rubygems/core_ext/kernel_require.rb:72:in `require'
> 	from /Users/tylervannurden/.rvm/rubies/ruby-2.6.5/lib/ruby/site_ruby/2.6.0/rubygems/core_ext/kernel_require.rb:72:in `require'
> 	from -e:1:in `<main>'
> ```

Make sure you have all the correct keys.

**Comment by tvannurden** [comment]
> > What are all these other changes that leaked in? The model annotations
> > .
> 
> This is because of changing from utf8 to utf8mb4 in database.yml. It was explained on slack and has it's own PR

If they're in a different PR then the changes need to stay in that PR, and not leak into this PR. Please remove

**Comment by tvannurden** [comment]
@maciejkocylapc you also need to limit how many records a client can receive. If I put `per=1,000,000`, your API will try and return 1 million records which is definitely not what we want.

**Review by kbiala — APPROVED** [review]

**Pull request merged**